lib/teamsnap.rb

```diff
@@ -131,18 +136,50 @@ def init(opts = {})
       apply_endpoints(self, collection) && true
     end
 
-    def run(via, href, args = {})
+    def run(via, href, args = {}, opts = {})
       resp = client.send(via, href, args)
 
       if resp.status == 200
-        Oj.load(resp.body)
-          .fetch(:collection)
+        collection = Oj.load(resp.body).fetch(:collection)
+        TeamSnap.write_backup_file(opts[:backup_cache_file], collection)
+        collection
       else
-        error_message = parse_error(resp)
-        raise TeamSnap::Error.new(error_message)
+        if TeamSnap.backup_file_exists?(opts[:backup_cache_file])
+          warn("Connection to API failed.. using backup cache file to initialize endpoints")
+          Oj.load(IO.read(opts[:backup_cache_file]))
+        else
+          error_message = parse_error(resp)
+          raise TeamSnap::Error.new(error_message)
+        end
+      end
+    end
+
+    def backup_file_for(backup_cache)
+      case backup_cache
+      when true
+        "./tmp/.teamsnap_rb"
+      else
+        backup_cache
+      end
+    end
+
+    def write_backup_file(file_location, collection)
+      return unless file_location
+      dir_location = File.dirname(file_location)
+      if Dir.exist?(dir_location)
+        File.open(file_location, "w+") { |f| f.write Oj.dump(collection) }
+      else
+        warn(
+          "WARNING: Directory '#{dir_location}' does not exist. " +
+          "Backup cache functionality will not work until this is resolved."
+        )
       end
     end
 
+    def backup_file_exists?(backup_cache_file)
+      !backup_cache_file.nil? && File.exist?(backup_cache_file)
```

Comment:
switched up... but interestingly: !!var is faster with strings.. !var.nil? is faster on nils